File: VirtualSense/DJ_VirtualMachine.1.0/darjeeling.1.1/src/lib/virtualsense/javax/virtualsense/storage/Storage.java
/*
 * Storage.java
 * 
 * Copyright (c) 2008-2010 CSIRO, Delft University of Technology.
 * 
 * This file is part of Darjeeling.
 * 
 * Darjeeling is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Darjeeling is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with Darjeeling.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 
package javax.virtualsense.storage;

import java.lang.String;

/**
 * 
 * API for store on flash memory.
 * 
 * @author Matteo Dromedari
 *
 */
public class Storage
{
	/**
	 * Write variable on flash memory.
	 * @param name Name of variable to write;
	 * @param var Variable to write. 
	 * @return 0 - Successful write;
	 * 		  -1 - Write failed.
	 */
	public static short writeVar(String name, String str)
	{
		return _writeVar(name, name.length(), str, str.length());	
	}
	
	public static short writeVar(String name, short s)
	{
		return _writeVar(name, name.length(), s);
	}
	
	public static short writeVar(String name, int var)
	{
		return _writeVar(name, name.length(), var);
	}
	
	private static native short _writeVar(String name, int nameLen, String str, int strLen);
	private static native short _writeVar(String name, int nameLen, short s);
	private static native short _writeVar(String name, int nameLen, int var);
	
	
	/**
	 * Read to flash memory specified variable.
	 * @param name Name of variable to read;
	 * @return specified variable in byte array.
	 */
	public static int readInt(String name)
	{
		return _readInt(name, name.length());
	}
	
	public static short readShort(String name)
	{
		return _readShort(name, name.length());
	}
	
	public static String readString(String name)
	{
		return _readString(name, name.length());
	}
	
	private static native int _readInt(String name, int nameLen);
	private static native short _readShort(String name, int nameLen);
	private static native String _readString(String name, int nameLen);
	
	
	/**
	 * Write byte onto specified file. Reserves a free memory space for save a file. At the first invocation the memory is reserved
	 * and the first byte is write, the next bytes are added in append.	
	 * @param name File name;
	 * @param b Byte to write;
	 * @return 0 - Successful write;
	 * 		  -1 - Write failed.
	 */
	
	/**
	 * Read a byte from specified file.
	 * @param name Name of file.
	 * @return read byte.
	 */
}

File: VirtualSense/DJ_VirtualMachine.1.0/darjeeling.1.1/src/vm/platform/contiki/storage.c
/*
 *	storage.c
 *
 *  Copyright (c) 2011 DiSBeF, University of Urbino.
 *
 *	This file is part of VirtualSense.
 *
 *	VirtualSense is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	VirtualSense is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with VirtualSense.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Storage definitions.
 *
 * @author Matteo Dromedari
 *
 */



#include "storage.h"
#include "contiki.h"
#include "dev/board.h"
#include "eeprom.h"
#include "base_definitions.h"
#include "common/array.h"

#include "common/heap/heap.h"
#include "common/execution/execution.h"
#include "common/panic.h"





/*---------------------------------------------------------------------------*/

/* Operation on variable */
short write_var(unsigned char *name, int name_len, unsigned char *var, int var_len)
{
	eeprom_addr_t mem_pointer = APPS_EEPROM_FS_BASE;

	eeprom_init();
	eeprom_write(mem_pointer, name, name_len);
	eeprom_write((mem_pointer + name_len), var, var_len);

	return 0;
}

int16_t read_short(unsigned char *name, int name_len)
{
	eeprom_addr_t mem_pointer = APPS_EEPROM_FS_BASE;
	int i;

	eeprom_init();
	unsigned char n[name_len];	
	eeprom_read(mem_pointer, n, name_len);
	printf("C - nome letto: ");
	for(i = 0; i < name_len; i++)
	{
		printf("%c", n[i]);
	}
	printf(" ");

	unsigned char v[sizeof(int16_t)];
	eeprom_read(mem_pointer + name_len, v, sizeof(int16_t));

	int16_t *ret = &v;

	return *ret;
}

int32_t read_int(unsigned char *name, int name_len)
{
	eeprom_addr_t mem_pointer = APPS_EEPROM_FS_BASE;
	int i;

	eeprom_init();
	unsigned char n[name_len];	
	eeprom_read(mem_pointer, n, name_len);
	printf("C - nome letto: ");
	for(i = 0; i < name_len; i++)
	{
		printf("%c", n[i]);
	}
	printf(" ");

	unsigned char v[sizeof(int32_t)];
	eeprom_read(mem_pointer + name_len, v, sizeof(int32_t));

	int32_t *ret = &v; 

	return *ret;
}

char* read_String(unsigned char *name, int name_len)
{
	eeprom_addr_t mem_pointer = APPS_EEPROM_FS_BASE;
	int i;
	char *str;

	eeprom_init();
	unsigned char n[name_len];	
	eeprom_read(mem_pointer, n, name_len);
	printf("C - nome letto: ");
	for(i = 0; i < name_len; i++)
	{
		printf("%c", n[i]);
	}
	printf(" ");

	unsigned char v[7];
	eeprom_read(mem_pointer + name_len, v, 6);
	v[6] = '\0';

	str = dj_mem_alloc(strlen(v) + 1, dj_vm_getSysLibClassRuntimeId(dj_exec_getVM(), BASE_CDEF_java_lang_String));

	if(str != NULL)
	{
		strcpy(str, v);
	}

	return str;
}




//short write_var(char name[], short s);
//short write_var(char name[], int i);
//uint8_t write_var(char name[], long l);
//short write_var(char name[], uint8_t b[]);
//uint8_t *read_var(char name[]);


/* Operation on file */
//short write_file(char name[], uint8_t b);
//uint8_t read_file(char name[]);



/*---------------------------------------------------------------------------*/

